### PR TITLE
STY: Old formatting to f-strings (PEP-498)

### PR DESCRIPTION
**Pull request by jond01**
%, '{}'.format -> f'{}'
F-strings are supported from py3.6 which is nibabel's minimal python version (up-to-date).

- Was done with:
 https://github.com/ikamensh/flynt
- Excluding versioneer.py
- First commit changes were reviewed manually

### CODE REVIEW

File: nibabel/data.py [hunk 1]
```diff
@@ -359,10 +356,6 @@ def datasource_or_bomber(pkg_def, **options):
...
     else:
         pkg_name = 'data at ' + unix_relpath
-    msg = ('%(name)s is version %(pkg_version)s but we need '
-           'version >= %(req_version)s\n\n%(pkg_hint)s' %
-           dict(name=pkg_name,
-                pkg_version=ds.version,
-                req_version=version,
-                pkg_hint=pkg_hint))
+    msg = (f"{pkg_name} is version {ds.version} but we need "
+           f"version >= {version}\n\n{pkg_hint}")
```
Comment (effigies) [inline nibabel/data.py:360]:
```suggestion
    msg = f"{pkg_name} is version {ds.version} but we need version >= {version}\n\n{pkg_hint}"
```

File: nibabel/freesurfer/tests/test_io.py [hunk 2]
```diff
@@ -49,12 +49,12 @@ def _hash_file_content(fname):
...
 def test_geometry():
     """Test IO of .surf"""
-    surf_path = pjoin(data_path, "surf", "%s.%s" % ("lh", "inflated"))
+    surf_path = pjoin(data_path, "surf", f"{'lh'}.{'inflated'}")
     coords, faces = read_geometry(surf_path)
     assert 0 == faces.min()
     assert coords.shape[0] == faces.max() + 1
 
-    surf_path = pjoin(data_path, "surf", "%s.%s" % ("lh", "sphere"))
+    surf_path = pjoin(data_path, "surf", f"{'lh'}.{'sphere'}")
```
Comment (effigies) [inline nibabel/freesurfer/tests/test_io.py:57]:
```suggestion
    surf_path = pjoin(data_path, "surf", "lh.sphere")
```

File: nibabel/freesurfer/tests/test_io.py [hunk 3]
```diff
@@ -133,7 +132,7 @@ def test_quad_geometry():
 @freesurfer_test
 def test_morph_data():
     """Test IO of morphometry data file (eg. curvature)."""
-    curv_path = pjoin(data_path, "surf", "%s.%s" % ("lh", "curv"))
+    curv_path = pjoin(data_path, "surf", f"{'lh'}.{'curv'}")
```
Comment (effigies) [inline nibabel/freesurfer/tests/test_io.py:135]:
```suggestion
    curv_path = pjoin(data_path, "surf", "lh.curv")
```

File: nibabel/freesurfer/tests/test_io.py [hunk 4]
```diff
@@ -171,7 +170,7 @@ def test_annot():
     """Test IO of .annot against freesurfer example data."""
     annots = ['aparc', 'aparc.a2005s']
     for a in annots:
-        annot_path = pjoin(data_path, "label", "%s.%s.annot" % ("lh", a))
+        annot_path = pjoin(data_path, "label", f"{'lh'}.{a}.annot")
```
Comment (effigies) [inline nibabel/freesurfer/tests/test_io.py:173]:
```suggestion
        annot_path = pjoin(data_path, "label", f"lh.{a}.annot")
```

File: nibabel/funcs.py [hunk 5]
```diff
@@ -134,15 +134,13 @@ def concat_images(images, check_affines=True, axis=None):
     for i, img in enumerate(images):
         if len(img.shape) != n_dim:
             raise ValueError(
-                'Image {0} has {1} dimensions, image 0 has {2}'.format(
-                    i, len(img.shape), n_dim))
+                f'Image {i} has {len(img.shape)} dimensions, image 0 has {n_dim}')
```
Comment (effigies) [inline nibabel/funcs.py:137]: Can be on the same line as `ValueError` now.

File: nibabel/funcs.py [hunk 6]
```diff
@@ -134,15 +134,13 @@ def concat_images(images, check_affines=True, axis=None):
...
-            raise ValueError('shape {0} for image {1} not compatible with '
-                             'first image shape {2} with axis == {3}'.format(
-                                 img.shape, i, shape0, axis))
+            raise ValueError(f'shape {img.shape} for image {i} not compatible with '
+                             f'first image shape {shape0} with axis == {axis}')
         if check_affines and not np.all(img.affine == affine):
-            raise ValueError('Affine for image {0} does not match affine '
-                             'for first image'.format(i))
+            raise ValueError(f'Affine for image {i} does not match affine for '
+                             f'first image')
```
Comment (effigies) [inline nibabel/funcs.py:143]:
```suggestion
            raise ValueError(f'Affine for image {i} does not match affine for first image')
```

File: nibabel/gifti/gifti.py [hunk 7]
```diff
@@ -487,7 +486,7 @@ def to_xml_open(self):
 \tExternalFileOffset="%d">\n"""
         di = ""
         for i, n in enumerate(self.dims):
-            di = di + '\tDim%s=\"%s\"\n' % (str(i), str(n))
+            di = di + f'\tDim{str(i)}="{str(n)}\"\n'
```
Comment (effigies) [inline nibabel/gifti/gifti.py:489]:
```suggestion
            di = di + f'\tDim{i}="{n}"\n'
```

File: nibabel/gifti/parse_gifti_fast.py [hunk 8]
```diff
@@ -177,7 +177,7 @@ def StartElementHandler(self, name, attrs):
                     attrs["ArrayIndexingOrder"]]
             num_dim = int(attrs.get("Dimensionality", 0))
             for i in range(num_dim):
-                di = "Dim%s" % str(i)
+                di = f"Dim{str(i)}"
```
Comment (effigies) [inline nibabel/gifti/parse_gifti_fast.py:180]:
```suggestion
                di = f"Dim{i}"
```

File: nibabel/gifti/tests/test_parse_gifti_fast.py [hunk 9]
```diff
@@ -106,8 +106,7 @@ def assert_default_types(loaded):
             continue
         loadedtype = type(getattr(loaded, attr))
         assert loadedtype == defaulttype, (
-            "Type mismatch for attribute: {} ({!s} != {!s})".format(
-                attr, loadedtype, defaulttype))
+            f"Type mismatch for attribute: {attr} ({loadedtype!s} != {defaulttype!s})")
```
Comment (effigies) [inline nibabel/gifti/tests/test_parse_gifti_fast.py:109]:
```suggestion
            f"Type mismatch for attribute: {attr} ({loadedtype} != {defaulttype})")
```

File: nibabel/keywordonly.py [hunk 10]
```diff
@@ -19,8 +19,8 @@ def decorator(func):
         def wrapper(*args, **kwargs):
             if len(args) > n:
                 raise TypeError(
-                    '{0} takes at most {1} positional argument{2}'.format(
-                        func.__name__, n, 's' if n > 1 else ''))
+                    f"{func.__name__} takes at most {n} positional "
+                    f"argument{'s' if n > 1 else ''}")
```
Comment (effigies) [inline nibabel/keywordonly.py:23]:
```suggestion
                    f"{func.__name__} takes at most {n} positional argument{'s' if n > 1 else ''}")
```

File: nibabel/nicom/dicomreaders.py [hunk 11]
```diff
@@ -95,12 +95,10 @@ def read_mosaic_dir(dicom_path,
...
-                raise DicomReadError('Could not find diffusion '
-                                     'information reading file "%s"; '
-                                     ' is it possible this is not '
-                                     'a _raw_ diffusion directory? '
-                                     'Could it be a processed dataset '
-                                     'like ADC etc?' % fname)
+                raise DicomReadError(
+                    f'Could not find diffusion information reading file "{fname}";  '
+                    f'is it possible this is not a _raw_ diffusion directory? '
+                    f'Could it be a processed dataset like ADC etc?')
```
Comment (effigies) [inline nibabel/nicom/dicomreaders.py:101]:
```suggestion
                    'is it possible this is not a _raw_ diffusion directory? '
                    'Could it be a processed dataset like ADC etc?')
```

File: nibabel/nifti1.py [hunk 12]
```diff
@@ -438,8 +438,8 @@ def __init__(self, code, content, parent_hdr=None):
             self._is_implicit_VR = False
             self._content = pdcm.dataset.Dataset()
         else:
-            raise TypeError("content must be either a bytestring or a pydicom "
-                            "Dataset. Got %s" % content.__class__)
+            raise TypeError(f"content must be either a bytestring or a "
+                            f"pydicom Dataset. Got {content.__class__}")
```
Comment (effigies) [inline nibabel/nifti1.py:442]:
```suggestion
            raise TypeError(f"content must be either a bytestring or a pydicom Dataset. "
                            f"Got {content.__class__}")
```

File: nibabel/nifti1.py [hunk 13]
```diff
@@ -519,7 +519,7 @@ def get_sizeondisk(self):
         return np.sum([e.get_sizeondisk() for e in self])
 
     def __repr__(self):
-        s = "Nifti1Extensions(%s)" % ', '.join(str(e) for e in self)
+        s = f"Nifti1Extensions({', '.join(str(e) for e in self)})"
         return s
```
Comment (effigies) [inline nibabel/nifti1.py:523]: This is harder to read. I would suggest we revert. Not sure why this was a two-step method, though...

```suggestion
        return "Nifti1Extensions(%s)" % ', '.join(str(e) for e in self)
```

File: nibabel/nifti1.py [hunk 14]
```diff
@@ -711,8 +711,8 @@ def write_to(self, fileobj):
                 self._structarr['vox_offset'] = min_vox_offset
             elif vox_offset < min_vox_offset:
                 raise HeaderDataError(
-                    'vox offset set to {0}, but need at least {1}'.format(
-                        vox_offset, min_vox_offset))
+                    f'vox offset set to {vox_offset}, but need at least '
+                    f'{min_vox_offset}')
```
Comment (effigies) [inline nibabel/nifti1.py:715]:
```suggestion
                    f'vox offset set to {vox_offset}, but need at least {min_vox_offset}')
```

File: nibabel/nifti1.py [hunk 15]
```diff
@@ -875,8 +875,8 @@ def set_data_shape(self, shape):
             else:
                 overflow = hdr['glmin'] != shape[0]
             if overflow:
-                raise HeaderDataError('shape[0] %s does not fit in glmax '
-                                      'datatype' % shape[0])
+                raise HeaderDataError(f'shape[0] {shape[0]} does '
+                                      f'not fit in glmax datatype')
```
Comment (effigies) [inline nibabel/nifti1.py:879]:
```suggestion
                raise HeaderDataError(f'shape[0] {shape[0]} does not fit in glmax datatype')
```

File: nibabel/nifti1.py [hunk 16]
```diff
@@ -1169,7 +1169,7 @@ def get_slope_inter(self):
             return None, None
         if not np.isfinite(inter):
             raise HeaderDataError(
-                'Valid slope but invalid intercept {0}'.format(inter))
+                f'Valid slope but invalid intercept {inter}')
```
Comment (effigies) [inline nibabel/nifti1.py:1172]: Should fit on the same line as `HeaderDataError`.

File: nibabel/nifti1.py [hunk 17]
```diff
@@ -1582,13 +1580,12 @@ def set_slice_times(self, slice_times):
                 matching_labels.append(label)
 
         if not matching_labels:
-            raise HeaderDataError('slice ordering of %s fits '
-                                  'with no known scheme' % st_order)
+            raise HeaderDataError(f'slice ordering of {st_order} '
+                                  f'fits with no known scheme')
```
Comment (effigies) [inline nibabel/nifti1.py:1584]:
```suggestion
            raise HeaderDataError(f'slice ordering of {st_order} fits with no known scheme')
```

File: nibabel/nifti1.py [hunk 18]
```diff
@@ -1582,13 +1580,12 @@ def set_slice_times(self, slice_times):
...
         if len(matching_labels) > 1:
             warnings.warn(
-                'Multiple slice orders satisfy: %s. Choosing the first one'
-                % ', '.join(matching_labels)
-            )
+                f"Multiple slice orders satisfy: {', '.join(matching_labels)}. "
+                f"Choosing the first one")
```
Comment (effigies) [inline nibabel/nifti1.py:1588]:
```suggestion
                "Choosing the first one")
```

File: nibabel/nifti1.py [hunk 19]
```diff
@@ -1682,8 +1678,7 @@ def _chk_magic(hdr, fix=False):
         magic = hdr['magic'].item()
         if magic in (hdr.pair_magic, hdr.single_magic):
             return hdr, rep
-        rep.problem_msg = ('magic string "%s" is not valid' %
-                           asstr(magic))
+        rep.problem_msg = (f'magic string "{asstr(magic)}" is not valid')
```
Comment (effigies) [inline nibabel/nifti1.py:1681]:
```suggestion
        rep.problem_msg = f'magic string "{asstr(magic)}" is not valid'
```

File: nibabel/nifti1.py [hunk 20]
```diff
@@ -1703,15 +1698,14 @@ def _chk_offset(hdr, fix=False):
...
+                rep.fix_msg = f'setting to minimum value of {hdr.single_vox_offset}'
             return hdr, rep
         if not offset % 16:
             return hdr, rep
         # SPM uses memory mapping to read the data, and
         # apparently this has to start on 16 byte boundaries
-        rep.problem_msg = ('vox offset (={0:g}) not divisible '
-                           'by 16, not SPM compatible'.format(offset))
+        rep.problem_msg = (f'vox offset (={offset:g}) not divisible '
+                           f'by 16, not SPM compatible')
```
Comment (effigies) [inline nibabel/nifti1.py:1708]:
```suggestion
        rep.problem_msg = f'vox offset (={offset:g}) not divisible by 16, not SPM compatible'
```

File: nibabel/openers.py [hunk 21]
```diff
@@ -24,8 +24,8 @@
 
     # < 0.7 - no good
     if StrictVersion(version) < StrictVersion('0.7.0'):
-        warnings.warn('indexed_gzip is present, but too old '
-                      '(>= 0.7.0 required): {})'.format(version))
+        warnings.warn(f'indexed_gzip is present, but too old '
+                      f'(>= 0.7.0 required): {version})')
```
Comment (effigies) [inline nibabel/openers.py:28]:
```suggestion
        warnings.warn(f'indexed_gzip is present, but too old (>= 0.7.0 required): {version})')
```

File: nibabel/optpkg.py [hunk 22]
```diff
@@ -99,18 +99,17 @@ def optional_package(name, trip_msg=None, min_version=None):
         # Failed version check
         if trip_msg is None:
             if callable(min_version):
-                trip_msg = 'Package %s fails version check' % min_version
+                trip_msg = f'Package {min_version} fails version check'
             else:
-                trip_msg = ('These functions need %s version >= %s' %
-                            (name, min_version))
+                trip_msg = (f'These functions need {name} '
+                            f'version >= {min_version}')
```
Comment (effigies) [inline nibabel/optpkg.py:105]:
```suggestion
                trip_msg = f'These functions need {name} version >= {min_version}'
```

File: nibabel/orientations.py [hunk 23]
```diff
@@ -336,10 +335,10 @@ def axcodes2ornt(axcodes, labels=None):
     labels = list(zip('LPI', 'RAS')) if labels is None else labels
     allowed_labels = sum([list(L) for L in labels], []) + [None]
     if len(allowed_labels) != len(set(allowed_labels)):
-        raise ValueError('Duplicate labels in {}'.format(allowed_labels))
+        raise ValueError(f'Duplicate labels in {allowed_labels}')
     if not set(axcodes).issubset(allowed_labels):
-        raise ValueError('Not all axis codes {} in label set {}'
-                         .format(list(axcodes), allowed_labels))
+        raise ValueError(f'Not all axis codes {list(axcodes)} '
+                         f'in label set {allowed_labels}')
```
Comment (effigies) [inline nibabel/orientations.py:341]:
```suggestion
        raise ValueError(f'Not all axis codes {list(axcodes)} in label set {allowed_labels}')
```

File: nibabel/parrec.py [hunk 24]
```diff
@@ -430,8 +430,7 @@ def vol_is_full(slice_nos, slice_max, slice_min=1):
     slice_set = set(range(slice_min, slice_max + 1))
     if not slice_set.issuperset(slice_nos):
         raise ValueError(
-            'Slice numbers outside inclusive range {0} to {1}'.format(
-                slice_min, slice_max))
+            f'Slice numbers outside inclusive range {slice_min} to {slice_max}')
```
Comment (effigies) [inline nibabel/parrec.py:433]: Can go on same line as `ValueError`.

File: nibabel/parrec.py [hunk 25]
```diff
@@ -733,9 +732,8 @@ def __init__(self, info, image_defs, permit_truncated=False,
         # dtype
         bitpix = self._get_unique_image_prop('image pixel size')
         if bitpix not in (8, 16):
-            raise PARRECError('Only 8- and 16-bit data supported (not %s)'
-                              'please report this to the nibabel developers'
-                              % bitpix)
+            raise PARRECError(f'Only 8- and 16-bit data supported (not {bitpix}) '
+                              f'please report this to the nibabel developers')
```
Comment (effigies) [inline nibabel/parrec.py:736]:
```suggestion
            raise PARRECError(f'Only 8- and 16-bit data supported (not {bitpix}) '
                              'please report this to the nibabel developers')
```

File: nibabel/spatialimages.py [hunk 26]
```diff
@@ -516,11 +516,11 @@ def __str__(self):
         shape = self.shape
         affine = self.affine
         return '\n'.join((str(self.__class__),
-                          'data shape %s' % (shape,),
+                          f'data shape {shape}',
                           'affine: ',
-                          '%s' % affine,
+                          f'{affine}',
```
Comment (effigies) [inline nibabel/spatialimages.py:521]: No advantage to f-strings here.

```suggestion
                          str(affine),
```

Unless you want to do something like combining the two lines: `'affine:\n{affine}'`

File: nibabel/spatialimages.py [hunk 27]
```diff
@@ -516,11 +516,11 @@ def __str__(self):
...
                           'metadata:',
-                          '%s' % self._header))
+                          f'{self._header}'))
```
Comment (effigies) [inline nibabel/spatialimages.py:523]: Same here.

```suggestion
                          str(self._header)))
```

Or `'metadata:\n{self._header}'))`.

File: nibabel/cmdline/dicomfs.py [hunk 28]
```diff
@@ -71,31 +71,31 @@ def get_paths(self):
...
             pd = paths.setdefault(study.patient_name_or_uid(), {})
             patient_info = 'patient information\n'
-            patient_info = 'name: %s\n' % study.patient_name
-            patient_info += 'ID: %s\n' % study.patient_id
-            patient_info += 'birth date: %s\n' % study.patient_birth_date
-            patient_info += 'sex: %s\n' % study.patient_sex
+            patient_info = f'name: {study.patient_name}\n'
+            patient_info += f'ID: {study.patient_id}\n'
+            patient_info += f'birth date: {study.patient_birth_date}\n'
+            patient_info += f'sex: {study.patient_sex}\n'
```
[the remark this answers is not in the document]
Comment (jond01) [reply nibabel/cmdline/dicomfs.py:77]: I think I'll skip it this time, anyway the dicom part in nibabel is not so mature

File: nibabel/cmdline/dicomfs.py [hunk 29]
```diff
@@ -196,8 +196,8 @@ def release(self, path, flags, fh):
 def get_opt_parser():
     # use module docstring for help output
     p = OptionParser(
-        usage="%s [OPTIONS] <DIRECTORY CONTAINING DICOMSs> <mount point>"
-              % os.path.basename(sys.argv[0]),
+        usage=f"{os.path.basename(sys.argv[0])} [OPTIONS] "
+              f"<DIRECTORY CONTAINING DICOMSs> <mount point>",
```
[the remark this answers is not in the document]
Comment (jond01) [reply nibabel/cmdline/dicomfs.py]: Ok, I agree

File: nibabel/streamlines/tck.py [hunk 30]
```diff
@@ -211,15 +211,15 @@ def save(self, fileobj):
             data_for_streamline = first_item.data_for_streamline
             if len(data_for_streamline) > 0:
                 keys = ", ".join(data_for_streamline.keys())
-                msg = ("TCK format does not support saving additional data"
-                       " alongside streamlines. Dropping: {}".format(keys))
+                msg = (f"TCK format does not support saving additional "
```
Comment (jond01) [inline nibabel/streamlines/tck.py:214]: Changed here

File: nibabel/streamlines/tck.py [hunk 31]
```diff
@@ -330,20 +330,20 @@ def _read_header(fileobj):
             hdr['datatype'] = "Float32LE"
 
         if not hdr['datatype'].startswith('Float32'):
-            msg = ("TCK only supports float32 dtype but 'datatype: {}' was"
-                   " specified in the header.").format(hdr['datatype'])
+            msg = (f"TCK only supports float32 dtype but 'datatype: "
```
Comment (jond01) [inline nibabel/streamlines/tck.py:333]: And here

File: nibabel/streamlines/tck.py [hunk 32]
```diff
@@ -211,15 +211,15 @@ def save(self, fileobj):
...
+                msg = (f"TCK format does not support saving additional "
+                       f"data alongside streamlines. Dropping: {keys}")
                 warnings.warn(msg, DataWarning)
 
             data_for_points = first_item.data_for_points
             if len(data_for_points) > 0:
                 keys = ", ".join(data_for_points.keys())
-                msg = ("TCK format does not support saving additional data"
-                       " alongside points. Dropping: {}".format(keys))
+                msg = (f"TCK format does not support saving additional "
```
Comment (jond01) [inline nibabel/streamlines/tck.py:221]: Also here

File: nibabel/streamlines/tractogram.py [hunk 33]
```diff
@@ -140,10 +140,9 @@ def extend(self, other):
         """
         if (len(self) > 0 and len(other) > 0 and
                 sorted(self.keys()) != sorted(other.keys())):
-            msg = ("Entry mismatched between the two PerArrayDict objects."
-                   " This PerArrayDict contains '{0}' whereas the other "
-                   " contains '{1}'.").format(sorted(self.keys()),
-                                              sorted(other.keys()))
+            msg = (f"Entry mismatched between the two PerArrayDict objects. "
```
Comment (jond01) [inline nibabel/streamlines/tractogram.py:143]: +1

File: nibabel/streamlines/tractogram.py [hunk 34]
```diff
@@ -365,8 +364,8 @@ def affine_to_rasmm(self, value):
         if value is not None:
             value = np.array(value)
             if value.shape != (4, 4):
-                msg = ("Affine matrix has a shape of (4, 4) but a ndarray with"
-                       "shape {} was provided instead.").format(value.shape)
+                msg = (f"Affine matrix has a shape of (4, 4) but a ndarray with"
```
Comment (jond01) [inline nibabel/streamlines/tractogram.py:367]: +1

**Comment by jond01** [comment]
I cannot exactly understand CI failure:
https://travis-ci.com/github/nipy/nibabel/jobs/352352424

**Comment by effigies** [comment]
Travis sometimes glitches. I've restarted the job. We'll see if it repeats.

File: nibabel/tests/test_deprecator.py [hunk 35]
```diff
@@ -94,28 +94,24 @@ def test_dep_func(self):
             assert func() is None
             assert len(w) == 1
         assert (func.__doc__ ==
-                    'foo\n\n* Will raise {} as of version: 99.4\n'
-                    .format(ExpiredDeprecationError))
+                f'foo\n\n* Will raise {ExpiredDeprecationError} '
+                'as of version: 99.4\n')
```
Comment (effigies) [inline nibabel/tests/test_deprecator.py:98]:
```suggestion
                f'foo\n\n* Will raise {ExpiredDeprecationError} as of version: 99.4\n')
```

File: nibabel/tests/test_volumeutils.py [hunk 36]
```diff
@@ -1224,9 +1224,8 @@ def read(self, n_bytes):
         array_from_file(shape, np.int8, NoStringIO())
     except IOError as err:
         message = str(err)
-    assert message == 'Expected {0} bytes, got {1} bytes from {2}\n' \
-                      ' - could the file be damaged?'.format(
-                      11390625000000000000, 0, 'object')
+    assert message == (f"Expected {11390625000000000000} bytes, got {0} "
+                       f"bytes from {'object'}\n - could the file be damaged?")
```
Comment (effigies) [inline nibabel/tests/test_volumeutils.py:1228]:
```suggestion
    assert message == ("Expected 11390625000000000000 bytes, got 0 "
                       "bytes from object\n - could the file be damaged?")
```

**Comment by effigies** [comment]
Thanks!

**Pull request merged**